packages/office-ui-fabric-react/src/utilities/keytips/KeytipManager.ts

```diff
@@ -0,0 +1,430 @@
+import { IKeytipProps } from '../../Keytip';
+import {
+  IKeySequence,
+  convertSequencesToKeytipID,
+  transitionKeysContain,
+  IKeytipTransitionKey,
+  fullKeySequencesAreEqual,
+  getDocument,
+  replaceElement,
+  findIndex,
+  mergeOverflowKeySequences,
+  Async
+} from '../../Utilities';
+import { constructKeytipExecuteTargetFromId } from './KeytipUtils';
+
+export interface IUniqueKeytip {
+  uniqueID: string;
+  keytip: IKeytipProps;
+}
+
+export class KeytipManager {
+  private static _instance: KeytipManager = new KeytipManager();
+
+  public keytipTree: KeytipTree = new KeytipTree();
+  public currentSequence: IKeySequence;
+  public keytips: IUniqueKeytip[] = [];
+
+  private _layer: KeytipLayerBase;
+
+  private _enableSequences: IKeytipTransitionKey[];
+  private _exitSequences: IKeytipTransitionKey[];
+  private _returnSequences: IKeytipTransitionKey[];
+  private _newCurrentKeytipSequences?: IKeySequence[];
+
+  private _keytipIDCounter = 0;
+
+  private _delayedKeytipQueue: string[] = [];
+  private _delayedQueueTimeout: number;
+  private _async: Async = new Async();
+
+  /**
+   * Static function to get singleton KeytipManager instance
+   *
+   * @returns {KeytipManager} Singleton KeytipManager instance
+   */
+  public static getInstance(): KeytipManager {
+    return this._instance;
+  }
+
+  /**
+   * Sets the _layer property and other related variables
+   * Should be called from the KeytipLayer constructor
+   *
+   * @param layer - KeytipLayerBase object
+   */
+  public init(layer: KeytipLayerBase): void {
+    this._layer = layer;
+    this.currentSequence = '';
+    // All guaranteed to be set because of defaultProps in KeytipLayer
+    this._enableSequences = this._layer.props.keytipStartSequences!;
+    this._exitSequences = this._layer.props.keytipExitSequences!;
+    this._returnSequences = this._layer.props.keytipReturnSequences!;
+  }
+
+  /**
+   * Registers a keytip
+   *
+   * @param keytipProps - Keytip to register
+   * @returns {string} Unique ID for this keytip
+   */
+  public registerKeytip(keytipProps: IKeytipProps): string {
+    const keytipTree = this.keytipTree;
+
+    // Create a unique keytip
+    const uniqueKeytip: IUniqueKeytip = this._constructUniqueKeytip(keytipProps);
+
+    // Add keytip to the Manager, Tree, and Layer
+    // Check if trying to register a duplicate keytip, if so just update
+    const keytipIndex = this._findKeytipIndex(uniqueKeytip);
+    if (keytipIndex < 0) {
+      // Add
+      this.keytips.push(uniqueKeytip);
+      keytipTree.addNode(keytipProps, uniqueKeytip.uniqueID);
+      this._layer && this._layer.setKeytips(this.getKeytips());
+
+      // Add the keytip to the queue to show later
+      if (this.keytipTree.isCurrentKeytipParent(keytipProps)) {
+        this._addKeytipToQueue(convertSequencesToKeytipID(keytipProps.keySequences));
+      }
+    } else {
+      // Update
+      this._updateKeytip(uniqueKeytip, keytipIndex);
```

Comment:
>keytipIndex [](start = 39, length = 11)

keytipIndex is not needed here is it?
public updateKeytip(keytipProps: IKeytipProps, uniqueID: string)

Reply:
never mind i was looking at wrong method 

---
In reply to: [180579029](https://github.com/OfficeDev/office-ui-fabric-react/pull/4462#discussion_r180579029) [](ancestors = 180579029)